File: python/210.py
"""Project Euler Problem 210: Obtuse Angled Triangles.

Find the number of points (x, y) satisfying |x|+|y|≤N such that (0, 0),
(N/4, N/4), and (x, y) form an obtuse triangle.
"""

from __future__ import annotations

import math


def sq(n: int) -> int:
    """Return n squared."""
    return n * n


def fsq(n: float) -> float:
    """Return n squared."""
    return n * n


def solve() -> int:
    """Solve Problem 210."""
    N = 10**9

    ans = 3 * sq(N) // 2 + sq(N // 4 + 1) - (N // 4 + 1) - 2

    sqrt2 = math.sqrt(2)
    min_x = int(-(N / 8 * (sqrt2 - 1)))

    for x in range(min_x, 0):
        dy = math.sqrt(2 * sq(N // 8) - fsq(N / 8 - x))
        ans += 4 * (2 * int(math.ceil(dy)) - 1)

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/210.c
/*
 * Project Euler 210: Obtuse Angled Triangles
 *
 * Count lattice points (x,y) with |x|+|y| <= N such that
 * (0,0), (N/4, N/4), (x,y) form an obtuse triangle.
 */
#include <stdio.h>
#include <math.h>

int main(void) {
    long long N = 1000000000LL;
    long long q = N / 4;

    /* Base count from the diamond minus the circle region */
    long long ans = 3LL * N / 2 * N + (q + 1) * (q + 1) - (q + 1) - 2;

    double sqrt2 = sqrt(2.0);
    long long r = N / 8;
    int min_x = (int)(-(double)r * (sqrt2 - 1.0));

    for (int x = min_x; x < 0; x++) {
        double dy = sqrt(2.0 * (double)r * (double)r - ((double)r - (double)x) * ((double)r - (double)x));
        int idy = (int)ceil(dy);
        ans += 4LL * (2 * idy - 1);
    }

    printf("%lld\n", ans);
    return 0;
}
